### PR TITLE
Plan API

### PR DESCRIPTION
**Pull request by koenvanderveen**
## Description
Aims to implement #5224 

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: tests/syft/core/node/common/plan/plan_test.py [hunk 1]
```diff
@@ -203,7 +209,118 @@ def test_plan_batched_execution() -> None:
...
+    local_model = MLP(th)  # type: ignore
+
+    @make_plan
+    def train(  # type: ignore
+        xs=th.rand([64 * 3, 1, 28, 28]),
+        ys=th.randint(0, 10, [64 * 3, 10]),
+        params=List(local_model.parameters()),
+    ):
+
+        model = local_model.send(ROOT_CLIENT)
```
Comment (iamtrask) [inline tests/syft/core/node/common/plan/plan_test.py:301]: local_model should be passed in as an argument, no? Otherwise the plan might never be able to find it when the plan is used in the wild somewhere.

File: tests/syft/core/node/common/plan/plan_test.py [hunk 2]
```diff
@@ -203,7 +209,118 @@ def test_plan_batched_execution() -> None:
...
+    alice = sy.VirtualMachine(name="alice")
+    alice_client = alice.get_root_client()
+
+    @make_plan
+    def mul_plan(inp=th.zeros((3)), inp2=th.zeros((3))) -> th.Tensor:  # type: ignore
+        return inp * inp2
+
+    t1, t2 = th.tensor([1, 2, 3]), th.tensor([1, 2, 3])
+    plan_pointer = mul_plan.send(alice_client)
+    res = plan_pointer(inp=t1, inp2=t2)
```
Comment (iamtrask) [inline tests/syft/core/node/common/plan/plan_test.py:242]: can this plan be run multiple times? let's add a test to be sure.

**Review by iamtrask — CHANGES_REQUESTED** [review]
Really excellent progress! A couple questions in the tests.

File: tests/syft/core/node/common/plan/plan_test.py (hunk 2 above)
Comment (koenvanderveen) [reply tests/syft/core/node/common/plan/plan_test.py:242]: Yes it can, there is already a nb in this pr doing that. Will add a test

File: tests/syft/core/node/common/plan/plan_test.py (hunk 1 above)
Comment (koenvanderveen) [reply tests/syft/core/node/common/plan/plan_test.py:301]: Yes that would be better. However, I ran into problems with sending sy.modules, by passing them as arguments, because it relies on the object2proto function, which is not implemented for sy.modules

Comment (koenvanderveen) [reply tests/syft/core/node/common/plan/plan_test.py:301]: So i think one of my next projects might be to make it serializable. If you have any ideas here let me know

**Comment by koenvanderveen** [comment]
@madhavajay what do you need from me to wrap this up?

**Comment by madhavajay** [comment]
> @madhavajay what do you need from me to wrap this up?

Its high on the list sorry, just getting swamped.

**Comment by koenvanderveen** [comment]
> > @madhavajay what do you need from me to wrap this up?
> 
> Its high on the list sorry, just getting swamped.

np

**Comment by xutongye** [comment]
Great work. I pulled this PR and added a notebook to demo a use case with Opacus+Plan, #5333 . From the process I create the notebook using the Plan API, I feel it very necessary to support `for` loop in Plan. Without supporting to `for` loop, the Plan object created is very very large when I define a full training process as a Plan.

**Comment by koenvanderveen** [comment]
sry for the mess above, accidentally pushed to this branch -_-'

File: tests/syft/core/node/common/plan/plan_test.py (hunk 1 above)
Comment (madhavajay) [reply tests/syft/core/node/common/plan/plan_test.py]: We are working on serializable nn.Module and then a refactor of Sy.Module for this.

**Review by madhavajay — APPROVED** [review]
This is excellent @koenvanderveen! I have played around the with the notebooks and copied PlanBuilder out into the code and added a test.

I noticed a few small things which I think we should address separately to not hold this up and all the new PRs that are depending on it.

I think we can't meaningfully refactor the Plans API until we start using it anyway.
